### PR TITLE
fix: dynamic context mapping in Proxy Block

### LINKED ISSUE #8: Fix Dynamic Context Mapping in Proxy Block
# Fix Dynamic Context Mapping in Proxy Block

## Summary

The Proxy Block module currently has issues with dynamic context mapping that prevent context-aware blocks (like the Body field block) from rendering correctly. The context mapping form elements are not displayed via AJAX, and when they are present, the context options are not properly populated with environment-specific contexts.

## Current Behavior

- ✅ Blocks without context requirements (e.g., Admin Menu) render correctly
- ❌ Context mapping form elements are not shown when selecting context-aware blocks via AJAX
- ❌ Context dropdown options are empty/not populated when context mapping forms do appear
- ❌ No differentiation between Layout Builder and Block UI placement contexts
- ❌ No dynamic discovery of available contexts based on placement environment

## Expected Behavior

When a user selects a target block that requires contexts:

1. **AJAX Form Update**: Context mapping form elements should appear immediately after target block selection
2. **Dynamic Context Discovery**: Available context options should be populated based on:
   - **Layout Builder context**: entity being edited, view mode, layout sections
   - **Block UI context**: route contexts (node from URL, user from URL, taxonomy term, etc.)
3. **Environment-Aware Options**: Context dropdown should show different options depending on placement:
   - Layout Builder: `layout_builder.entity`, `layout_builder.view_mode`, route contexts
   - Block UI: route-based contexts like `@node.node_route_context`, `@user.current_user_context`
4. **Automatic Mapping**: When possible, automatically map contexts with matching names
5. **Validation**: Proper validation that all required contexts are mapped before form submission

### Current Implementation Issues

The current implementation in `ProxyBlock.php` has several problems:

1. **AJAX Structure Issue** (`buildTargetBlockConfigurationForm()` around line 230):
   - Context mapping form is built but may not be properly included in AJAX response
   - The form structure doesn't properly integrate context mapping with the AJAX wrapper

2. **Limited Context Discovery** (`buildContextMappingForm()` around line 304):
   - Only uses `$this->getContexts()` which provides limited proxy block contexts
   - Doesn't differentiate between Layout Builder and Block UI environments
   - Doesn't dynamically discover route-based contexts

3. **Static Context Options** (around line 313):
   - Context options are generated from proxy block contexts only
   - No environment-specific context discovery
   - Missing integration with Drupal's context system

#### 1. Fix AJAX Integration

The context mapping form needs to be properly integrated into the AJAX response structure:

```php
// In buildTargetBlockConfigurationForm()
if ($target_block instanceof ContextAwarePluginInterface) {
  $context_form = $this->buildContextMappingForm($target_block, $config);
  if (!empty($context_form)) {
    $form_elements['context_mapping'] = $context_form;
  }
}
```

The context mapping should be included **before** the target block's own configuration form, not after.

#### 2. Implement Dynamic Context Discovery

Create a new method to discover available contexts based on environment:

```php
/**
 * Discovers available contexts based on the current environment.
 *
 * @return array
 *   Array of context options keyed by context name.
 */
private function discoverAvailableContexts(): array {
  $contexts = [];
  
  // Always include proxy block contexts
  $proxy_contexts = $this->getContexts();
  
  // Detect if we're in Layout Builder
  if ($this->isLayoutBuilderEnvironment()) {
    $contexts += $this->getLayoutBuilderContexts();
  }
  
  // Add route-based contexts
  $contexts += $this->getRouteContexts();
  
  // Add entity contexts from current route
  $contexts += $this->getEntityContextsFromRoute();
  
  return $contexts;
}
```

#### 3. Environment Detection

Implement methods to detect the current placement environment:

```php
/**
 * Determines if the proxy block is being configured in Layout Builder.
 */
private function isLayoutBuilderEnvironment(): bool {
  // Check current route, request parameters, or other indicators
  // that suggest Layout Builder usage
}

/**
 * Gets Layout Builder specific contexts.
 */
private function getLayoutBuilderContexts(): array {
  // Return contexts like 'layout_builder.entity', 'layout_builder.view_mode'
}

/**
 * Gets route-based contexts.
 */
private function getRouteContexts(): array {
  // Discover contexts from current route parameters
}
```

#### 4. Enhanced Context Mapping Form

Update `buildContextMappingForm()` to use the dynamic context discovery:

```php
private function buildContextMappingForm(ContextAwarePluginInterface $target_block, array $config): array {
  $context_definitions = $target_block->getContextDefinitions();
  
  if (empty($context_definitions)) {
    return [];
  }

  // Use dynamic context discovery instead of static getContexts()
  $available_contexts = $this->discoverAvailableContexts();
  $context_options = ['' => $this->t('- Select context -')];
  
  foreach ($available_contexts as $context_name => $context) {
    $label = $context->getContextDefinition()->getLabel() ?: $context_name;
    $data_type = $context->getContextDefinition()->getDataType();
    $context_options[$context_name] = $this->t('@label (@type)', [
      '@label' => $label,
      '@type' => $data_type,
    ]);
  }

  // Rest of the form building logic...
}
```

#### Layout Builder Contexts
- `layout_builder.entity` - The entity being edited in Layout Builder
- `layout_builder.view_mode` - The view mode being edited
- `layout_builder.section` - Current layout section (if applicable)

#### Route-Based Contexts
- `@node.node_route_context` - Node from current route
- `@user.current_user_context` - Current user
- `@user.user_route_context` - User from route parameter
- `@taxonomy_term.taxonomy_term_route_context` - Term from route
- `@commerce_product.commerce_product_route_context` - Commerce product (if applicable)

#### Global Contexts
- Current user
- Current language
- Current theme
- Site configuration contexts

### AJAX Functionality
- [ ] Context mapping form appears immediately when selecting a context-aware target block
- [ ] Form elements update correctly via AJAX without page reload
- [ ] No JavaScript errors in browser console during AJAX operations

### Context Discovery
- [ ] Context options are populated in dropdown menus
- [ ] Different context options appear for Layout Builder vs Block UI placement
- [ ] Route-based contexts (node, user, term) are automatically discovered
- [ ] Context options show meaningful labels and data types

### Layout Builder Integration
- [ ] Layout Builder entity context is available when configuring blocks in Layout Builder
- [ ] View mode context is available for Layout Builder configurations
- [ ] Layout Builder contexts take precedence over route contexts when available

### Block UI Integration
- [ ] Route contexts (node from URL, user from URL) are available in Block UI
- [ ] Global contexts are available in both environments
- [ ] Context discovery works on entity pages (node/123, user/456, etc.)

### Validation and UX
- [ ] Required context mappings are clearly marked
- [ ] Form validation prevents submission with missing required contexts
- [ ] Clear error messages for context mapping issues
- [ ] Automatic context mapping when names match exactly

### Testing Requirements
- [ ] Unit tests for context discovery methods
- [ ] Kernel tests for context mapping functionality
- [ ] Functional tests for AJAX form behavior
- [ ] Functional JavaScript tests for Layout Builder integration

### Dependencies and Services

The implementation will likely need these additional services injected:

```php
/**
 * The current route match.
 */
protected RouteMatchInterface $routeMatch;

/**
 * The request stack.
 */
protected RequestStack $requestStack;

/**
 * The context repository.
 */
protected ContextRepositoryInterface $contextRepository;
```

### Backward Compatibility

- Existing proxy block configurations should continue to work
- Context mappings should gracefully handle missing contexts
- No breaking changes to the public API

### Performance Considerations

- Context discovery should be cached within the form building process
- Avoid expensive operations during AJAX callbacks
- Consider lazy loading of context options

### Security Considerations

- Validate that users have access to contexts they're trying to map
- Don't expose sensitive context information in form options
- Maintain existing access control for target blocks

## Related Files

- `src/Plugin/Block/ProxyBlock.php` - Main implementation file (lines 230-350 primary focus)
- `tests/src/Functional/` - Functional tests for form behavior
- `tests/src/FunctionalJavascript/` - JavaScript tests for AJAX functionality
- `tests/src/Kernel/` - Context mapping and discovery tests

### Manual Testing Steps

1. **Layout Builder Test**:
   - Create a content type with a Body field
   - Enable Layout Builder for the content type
   - Add a Proxy Block to the layout
   - Select "Body" as target block
   - Verify context mapping form appears
   - Verify entity context is available in dropdown
   - Save and verify the body content renders

2. **Block UI Test**:
   - Go to `/admin/structure/block`
   - Add a Proxy Block
   - Select "Body" as target block
   - Verify context mapping form appears  
   - Verify route-based contexts are available
   - Place block on a node page
   - Verify body content renders correctly

3. **AJAX Test**:
   - Configure a Proxy Block
   - Change target block selection multiple times
   - Verify context mapping form updates correctly each time
   - Verify no JavaScript errors occur

### Automated Testing

Create comprehensive test coverage including:

- Context discovery unit tests
- AJAX form functionality tests  
- Integration tests with Layout Builder
- Edge cases (missing contexts, invalid mappings)

## Priority

**High Priority** - This bug prevents the module from working correctly with context-aware blocks, which significantly limits its usefulness in real-world scenarios.

### PR DESCRIPTION
**Pull request by e0ipso**
This PR fixes the dynamic context mapping issues in the Proxy Block module by implementing a comprehensive context discovery system.

## Key Changes

1. **Dynamic Context Discovery System**:
   - Added RouteMatchInterface, RequestStack, and ContextRepositoryInterface services
   - Implemented environment-aware context discovery (Layout Builder vs Block UI)
   - Added support for Layout Builder contexts and route-based contexts

2. **Fixed AJAX Integration**:
   - Context mapping form now appears before target block configuration
   - Proper form structure ensures AJAX updates work correctly

3. **Enhanced Context Options**:
   - Context dropdowns populated with meaningful labels and data types
   - Graceful error handling for missing contexts

4. **Backward Compatibility**:
   - Existing configurations continue to work
   - No breaking changes to public API

Fixes #8

### CODE REVIEW

File: src/Plugin/Block/ProxyBlock.php [hunk 1]
```diff
@@ -290,6 +323,191 @@ private function buildTargetBlockConfigurationForm(string $plugin_id, FormStateI
...
+    // Check for Layout Builder routes.
+    $layout_builder_routes = [
+      'layout_builder.overrides.node.view',
+      'layout_builder.overrides.node.save',
+      'layout_builder.defaults.node.view',
+      'layout_builder.defaults.node.save',
+      'layout_builder.choose_block',
+      'layout_builder.add_block',
+      'layout_builder.update_block',
+    ];
```
Comment (e0ipso) [inline src/Plugin/Block/ProxyBlock.php:389]: @claude I don't like this list of hardcoded routes. Are you sure there isn't a dynamic way to detect this?

File: src/Plugin/Block/ProxyBlock.php [hunk 2]
```diff
@@ -290,6 +323,191 @@ private function buildTargetBlockConfigurationForm(string $plugin_id, FormStateI
...
+    $contexts = [];
+    
+    try {
+      // Get Layout Builder entity context.
+      $entity_contexts = $this->contextRepository->getRuntimeContexts(['@layout_builder.entity']);
+      $contexts += $entity_contexts;
+      
+      // Get Layout Builder view mode context.
+      $view_mode_contexts = $this->contextRepository->getRuntimeContexts(['@layout_builder.view_mode']);
+      $contexts += $view_mode_contexts;
```
Comment (e0ipso) [inline src/Plugin/Block/ProxyBlock.php:423]: @claude this seems fragile. Search for the place where Drupal core does this (likely in the layout_builder module) and ensure you do the same.

File: src/Plugin/Block/ProxyBlock.php [hunk 3]
```diff
@@ -290,6 +323,191 @@ private function buildTargetBlockConfigurationForm(string $plugin_id, FormStateI
...
+    $contexts = [];
+    
+    try {
+      // Get common route contexts.
+      $route_context_ids = [
+        '@node.node_route_context',
+        '@user.current_user_context',
+        '@user.user_route_context',
+        '@taxonomy_term.taxonomy_term_route_context',
+      ];
```
Comment (e0ipso) [inline src/Plugin/Block/ProxyBlock.php:450]: @claude again, this is a hard-coded list of paths, let's find a way to derive them dynamically.

File: src/Plugin/Block/ProxyBlock.php [hunk 4]
```diff
@@ -290,6 +323,191 @@ private function buildTargetBlockConfigurationForm(string $plugin_id, FormStateI
...
+   * @return array
+   *   Array of entity contexts from route.
+   */
+  private function getEntityContextsFromRoute(): array {
+    $contexts = [];
+    
+    // Get route parameters that might be entities.
+    $route_parameters = $this->routeMatch->getParameters()->all();
+    
+    foreach ($route_parameters as $parameter_name => $parameter_value) {
```
Comment (e0ipso) [inline src/Plugin/Block/ProxyBlock.php:484]: @claude I dislike foreach. I prefer `array_map` and friends.

File: src/Plugin/Block/ProxyBlock.php [hunk 5]
```diff
@@ -290,6 +323,191 @@ private function buildTargetBlockConfigurationForm(string $plugin_id, FormStateI
...
+      if (is_object($parameter_value) && method_exists($parameter_value, 'getEntityTypeId')) {
```
Comment (e0ipso) [inline src/Plugin/Block/ProxyBlock.php:485]: @claude this project favors guard clauses to simplify nesting, and maintainability.

File: src/Plugin/Block/ProxyBlock.php [hunk 6]
```diff
@@ -290,6 +323,191 @@ private function buildTargetBlockConfigurationForm(string $plugin_id, FormStateI
...
+    
+    if (in_array($route_name, $layout_builder_routes, TRUE)) {
+      return TRUE;
+    }
+    
+    // Check request for Layout Builder indicators.
+    $request = $this->requestStack->getCurrentRequest();
+    if ($request) {
+      $request_uri = $request->getRequestUri();
+      if (strpos($request_uri, 'layout_builder') !== FALSE) {
```
Comment (e0ipso) [inline src/Plugin/Block/ProxyBlock.php:399]: @claude is partial string-matching on request URIs a reliable way to accomplish this?

**Review by e0ipso — COMMENTED** [review]
@claude Take a look at all the PR review comments and address them properly. For each fix, take an additional moment to reflect if the fix needs to be applied elsewhere in the code base as well.
